keps/sig-scheduling/3902-decoupled-taint-manager/README.md

```diff
@@ -0,0 +1,300 @@
+- [ ]  Supporting documentation—e.g., additional design documents, links to mailing list discussions/SIG meetings, relevant PRs/issues, release notes
+
+## Summary
+
+In Kubernetes, `NodeLifecycleController` applies predefined `NoExecute` taints (e.g., `Unreachable`, `NotReady`) when the nodes are determined to be unhealthy. After the nodes get tainted, `TaintManager` does its due diligence to start deleting running pods on those nodes based on `NoExecute` taints, which can be added by anyone.
+
+In this KEP, we propose to 
+
+* Decouple `TaintManager` that performs taint-based pod eviction from `NodeLifecycleController` and make them two separate controllers: `NodeLifecycleManager` to add taints to unhealthy nodes and `TaintManager` to perform pod deletion on nodes tainted with NoExecute effect.
+
+* Allow users to opt-out from the taint-based eviction while still retaining the overall functionality provided by `NodeLifecycleController.`
+
+## Motivation
+* `NodeLifecycleController` combines two independent functions: adding a pre-defined set of `NoExecute` taints to unhealthy nodes and performing pod eviction on arbitrary `NoExecute` taints. While  `NodeLifecycleController`'s responsibility is primarily on Node unhealthy-zone, `TaintManager` is concerned with actioning the taints added by `NodeLifecycleController` by performing Pods eviction. De-coupling `NoExecuteTaintManager` from `NodeLifecycleController` in `kube-controller-manager` would allow to implement a single eviction controller to take care of general Pod eviction and deletion for different cases based on workload needs, including `NoExecute` taint-based pod eviction other than node `Unhealthy` and `Unreachable`.  Splitting `NodeLifecycleController` into two separate controllers will support more flexible extension and improvement of the existing node taint-based pod eviction. 
+* There are increasing needs to customize and extend the default node taint-based pod evictions in practice.
+  * Whether or not evict a stateful pod with local storage on `NoExecute` can depend on the actual taint conditions, workload status and other conditions. The decision may require information from the external systems and controllers, which have more knowledge and context about the workloads.  A custom `taint-manager`  will be able to control how and when a pod is evicted or deleted, including making calls to external systems and take actions on  `NoExecute` taints with other conditions such as unhealthy network. 
+  * Starting kubernetes 1.27, with the removal of the flag `enable-taint-manager`, replacing the default `NoExecute` taint-manager with a custom one would be more challenging. `NoExecute` taints can disable the no-execute taint manager, but such disabling can only be achieved through a mutating webhook on all pods, and it can interact negatively with the upstream `DefaultTolerations` webhook.
```

Comment:
This cannot be a motivation, as `enable-taint-manager=false` would still evict pods, just through a different (even less flexible) mechanism

Reply:
Yes, `descheduler` or other controllers can still evict pods, but `enable-taint-manager=false` will disable `NoExecute`-based pod eviction performed by the default `TaintManager`. We will remove it if you are strongly against it. 
 /cc @atosatto

Reply:
I'm not talking about descheduler. I'm talking about kube-controller-manager itself. When you set `enable-taint-manager=false`, kcm can evict pods just based on Node conditions.

Reply:
> When you set enable-taint-manager=false, kcm can evict pods just based on Node conditions.

@alculquicondor this is correct. However the functionality was deprecated and has been removed in https://github.com/kubernetes/kubernetes/pull/115840 as part of the effort to improve code-health in NodeLifeCycleController which this KEP is part of (as per your comment in https://github.com/kubernetes/enhancements/pull/3901#discussion_r11937486320).

The main benefit from de-coupling taint-based eviction from NodeLifeCycleController is that it would enable end users who have the possibility to control their control-plane configuration, to disable entirely taint-based eviction in their clusters without having to mutate all of the Pods via tolerations. 

While I do agree that there is room for extending the current built-in APIs to add (e.g. web-hooks) this should be part of a  different proposal and not in-scope of the current KEP which is purely focusing on re-organising the code.

Reply:
That is correct, it's removed.

But the KEP here is trying to say that because we removed the flag, there is less control over eviction. That is not true, thus can't be part of the motivation.

Reply:
> the main benefit from de-coupling taint-based eviction from NodeLifeCycleController is that it would enable end-users who have the possibility to control their control-plane configuration, to disable entirely taint-based eviction in their clusters without having to mutate all of the Pods via tolerations.

I disagree. Most end-users won't be able to disable the taint-based-eviction. And if they want to have some behavior similar to NoExecute taints with extra functionality, they still have to re-implement what kcm offers. So this is not a good extension mechanism.

> While I do agree that there is room for extending the current built-in APIs to add (e.g. web-hooks) this should be part of a different proposal and not in-scope of the current KEP which is purely focusing on re-organising the code.

I fully agree. That's what I have listed as one of the options to move forward with this KEP: to focus on re-organising.

Reply:
Updated the alternative with a focus on the code refactoring too. 

## Alternatives

Keeping the code as-is is an option, but it will make future extension of `TaintManager` harder and less flexible